policybot/cmd/server.go

```diff
@@ -115,3 +151,242 @@ func serverCmd() *cobra.Command {
+	router := mux.NewRouter()
+	handler = router
+	if a.StartupOptions.HTTPSOnly {
+		log.Infof("Using httpsOnly mode")
+		handler = httpsRedirectHandler(router)
+	}
+	// secret state for OAuth exchanges
+	secretState := make([]byte, 32)
+	if _, err := rand.Read(secretState); err != nil {
+		return fmt.Errorf("unable to generate secret state: %v", err)
+	}
+
+	s := &server{
+		listener: listener,
+		httpServer: http.Server{
+			Addr:           listener.Addr().(*net.TCPAddr).String(),
+			ReadTimeout:    10 * time.Second,
+			WriteTimeout:   10 * time.Second,
+			MaxHeaderBytes: 1 << 20,
+			Handler:        handler,
+		},
+		clientID:     a.StartupOptions.GitHubOAuthClientID,
+		clientSecret: a.StartupOptions.GitHubOAuthClientSecret,
+		secretState:  base64.StdEncoding.EncodeToString(secretState),
+		dashboard:    dashboard.New(router),
+	}
+
+	monitor, err := cfgmonitor.NewMonitor(context.Background(), ght, a.StartupOptions.ConfigRepo, a.StartupOptions.ConfigFile, s.Close)
+	if err != nil {
+		return fmt.Errorf("unable to create config monitor: %v", err)
+	}
+
+	// github webhook filters (keep refresher first in the list such that other filter see an up-to-date view in storage)
+	filters := []filters.Filter{
+		refresher.NewRefresher(context.Background(), cache, ght, a.Orgs),
+		nag,
+		labeler,
+		monitor,
+	}
+
+	ghHandler, err := githubwebhook.NewHandler(a.StartupOptions.GitHubWebhookSecret, filters...)
+	if err != nil {
+		return fmt.Errorf("unable to create GitHub webhook: %v", err)
+	}
+
+	// top-level handlers
+	router.Handle("/githubwebhook", ghHandler).Methods("POST")
+	router.Handle("/flakechaser", flakechaser.New(ght, store, cache, a.FlakeChaser)).Methods("GET")
+	router.Handle("/zenhubwebhook", zenhubwebhook.NewHandler(store, cache)).Methods("POST")
+	router.Handle("/sync", syncer.NewHandler(context.Background(), ght, cache, zht, store, bs, a.Orgs)).Methods("GET")
+	router.HandleFunc("/login", s.handleLogin)
+	router.HandleFunc("/githuboauthcallback", s.handleOAuthCallback)
+
+	// UI topics
+	s.dashboard.RegisterTopic(maintainers.NewTopic(store, cache))
+	s.dashboard.RegisterTopic(members.NewTopic(store, cache))
+	s.dashboard.RegisterTopic(issues.NewTopic(store, cache))
+	s.dashboard.RegisterTopic(pullrequests.NewTopic(store, cache))
+	s.dashboard.RegisterTopic(perf.NewTopic(store, cache))
+	s.dashboard.RegisterTopic(commithub.NewTopic(store, cache))
+	s.dashboard.RegisterTopic(flakes.NewTopic(store, cache))
+	s.dashboard.RegisterTopic(coverage.NewTopic(store, cache))
+	s.dashboard.RegisterTopic(features.NewTopic(store, cache))
+	s.dashboard.RegisterTopic(home.NewTopic(s.dashboard.Topics()))
+	s.dashboard.RegisterPageNotFound()
+
+	log.Infof("Listening on port %d", a.StartupOptions.Port)
+	err = s.httpServer.Serve(s.listener)
+	if err != http.ErrServerClosed {
+		return fmt.Errorf("listening on port %d failed: %v", a.StartupOptions.Port, err)
+	}
+
+	return nil
+}
+
+func (s *server) Close() {
+	if s.listener != nil {
```

Comment:
What will happen to some of the clients during close? Would it make sense to have a context that gets closed as part of server.Close?

Reply:
Actually, nobody calls Close and it was on my list to just take that method out.

Reply:
I take it back, Close is being called when config changes.

The clients aren't holding open any connections, so they will just be garbage collected. But it would be wise to have a generic cancellation story there so that any background go routines can get cleaned up, etc.

Reply:
I've updated use of context throughout the whole code base to be inline with expected use patterns for this type.